docs/proposals/13-seed-management.md

```diff
@@ -0,0 +1,91 @@
+
+In Gardener, scheduling shoots on seeds is quite similar to scheduling pods on nodes in Kubernetes. Therefore, a guiding principle behind the proposed design approaches is taking advantage of best practices and existing components already used in Kubernetes. 
+
+## Migrating Shoots Away from Unhealthy Seeds
+
+The first important automated seed management scenario is migrating shoots away from unhealthy seeds to healthy ones automatically or semi-automatically. This scenario is triggered by detecting a seed that for some reason can't function normally and should be either deleted or repaired "offline". Once such a seed is detected, it should be drained of all its shoots, which means that they should be evicted from the seed and scheduled to other, healthy seeds.
+
+A prerequisite for this scenario is [GEP 7: Shoot Control Plane Migration](07-shoot-control-plane-migration.md), including avoiding "split brain" situations using [leader election](07-shoot-control-plane-migration.md#leader-election-and-control-plane-termination).
+
+### Marking Seeds as "To Be Drained"
+
+The first step in migrating shoots away from an unhealthy seed is marking the seed as "to be drained" via a special `drain` annotation. Initially, this could simply be done manually, or perhaps assisted by a special `gardenctl` command, for example `gardenctl drain <seed>`, which simply adds the annotation to the specified seed. 
+
+Marking a seed as "to be drained" has the following effect:
+
+* It triggers the eviction of all shoots from this seed as described in [Evicting Shoots from Seeds Marked as "To be Drained"](#evicting-shoots-from-seeds-marked-as-to-be-drained) below.
+* It prevents any new shoots to be scheduled on this seed.
+
+Later, detecting unhealthy seeds and marking them as "to be drained" could also be done automatically. This could be achieved by introducing a new controller in `gardener-controller-manager` that observes seeds and reacts on certain seed conditions by adding the `drain` annotation mentioned above. Such conditions could include for example:
+
+* The seed is `NotReady` for a certain period of time.
+* A certain percentage of the seed's shoots are `NotReady` for a certain period of time.
+
+### Evicting Shoots from Seeds Marked as "To be Drained"
+
+The next step in migrating shoots away from an unhealthy seed is actually draining the seed that has been marked as described above, which involves evicting all shoots from this seed. This could be achieved by extending the `gardenlet` seed controller to react on the `drain` annotation by marking all seed shoots as `orphaned`, either by adding an `orphan` annotation or simply be deleting the `seed` field in the shoot specification.
```

Comment:
Hm, instead of working with annotations here, shouldn't we be inspired by Kubernetes (and the NPD, particularly) again and work with conditions? Based on them, similar to how MCM starts draining unhealthy nodes after a certain timeout, we could start draining seeds when relevant conditions are `False`?

Reply:
At least in the first discussion that we had, we were not sure that everything should be automated right from the start. This is why in the proposal there is a way to explicitly mark a seed as "to be drained", the assumption here being that the actual marking could be a manual action, and will mostly be a manual action at least in the beginning, this is why I chose to introduce a new annotation. Of course, this should be changed if it is decided that full automation should be pursued right from the start.

Reply:
Hmm, can we rather work with taints on the Seed just like in kubernetes (see [here](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/))?
E.g. adding a taint with effect `NoExecute` on a Seed would evict all Shoots on that Seed?

Reply:
@timebertt Updated the document to rely on a `NoSchedule` taint to mark the seed as "unschedulable", and described a behavior of the `gardenctl drain <seed>` command that is very similar to the behavior of the `kubectl drain <node>` command, which I consider the "best practice" to follow. `NoExecute` taints are not currently used as part of this process, i.e. they are a separate feature that does not play any part in draining a node. We could introduce such taints for seeds as well, it just goes a bit beyond my intended scope - unless of course we decide that supporting `NoExecute` taints has value in itself and then I could describe it as a separate "scenario" in the document.